SI-6260 Avoid double-def error with lambdas over value classes

retronym opened the pull request:
Post-erasure of value classs in method signatures to the underlying
type wreaks havoc when the erased signature overlaps with the
generic signature from an overriden method. There just isn't room
for both. But we _really_ need both; callers to the interface method
will be passing boxed values that the bridge needs to unbox and
pass to the specific method that accepts unboxed values.

This most commonly turns up with value classes that erase to
Object that are used as the parameter or the return type of
an anonymous function.

This was thought to have been intractable, unless we chose
a different name for the unboxed, specific method in the
subclass. But that sounds like a big task that would require
call-site rewriting, ala specialization.

But there is an important special case in which we don't need
to rewrite call sites. If the class defining the method is
anonymous, there is actually no need for the unboxed method;
it will _only_ ever be called via the generic method.

I came to this realisation when looking at how Java 8 lambdas
are handled. I was expecting bridge methods, but found none.
The lambda body is placed directly in a method exactly matching
the generic signature.

This commit detects the clash between bridge and target,
and recovers for anonymous classes by mangling the name
of the target method's symbol. This is used as the bytecode
name. The generic bridge forward to that, as before, with
the requisite box/unbox operations.

Review by @gkossakowski

retronym committed:
SI-6260 Avoid double-def error with lambdas over value classes

Post-erasure of value classs in method signatures to the underlying
type wreaks havoc when the erased signature overlaps with the
generic signature from an overriden method. There just isn't room
for both. But we *really* need both; callers to the interface method
will be passing boxed values that the bridge needs to unbox and
pass to the specific method that accepts unboxed values.

This most commonly turns up with value classes that erase to
Object that are used as the parameter or the return type of
an anonymous function.

This was thought to have been intractable, unless we chose
a different name for the unboxed, specific method in the
subclass. But that sounds like a big task that would require
call-site rewriting, ala specialization.

But there is an important special case in which we don't need
to rewrite call sites. If the class defining the method is
anonymous, there is actually no need for the unboxed method;
it will *only* ever be called via the generic method.

I came to this realisation when looking at how Java 8 lambdas
are handled. I was expecting bridge methods, but found none.
The lambda body is placed directly in a method exactly matching
the generic signature.

This commit detects the clash between bridge and target,
and recovers for anonymous classes by mangling the name
of the target method's symbol. This is used as the bytecode
name. The generic bridge forward to that, as before, with
the requisite box/unbox operations.

S11001001 commented:
Nice, works for me, thanks @retronym .

vendethiel commented on test/files/neg/t6260-named.scala:
```diff
@@ -0,0 +1,15 @@
+  (x: C[Any]) => {println(s"f($x)"); x} // okay
+  new T[C[Any]] { def apply(a: C[Any]) = a } // okay
+
+  // we can't rename the specific apply methid to avoid the clash
```
typo "method"

adriaanm commented:
Nice!

retronym commented:
I just realised that I should remove the bridge flag. Otherwise people java reflecting on the class might be confused.

gkossakowski commented:
@retronym: it's pretty awesome you've made that realization you describe! I missed the fact that anonymous classes has fewer constraints they need to respect.

Would it make sense to drop the bridge method in delambdafy phase altogether so we follow what Java 8 does? It would mean that you perform the whole unboxing business in `apply` method and never generate bridge method (name mangled or not). WDYT?

retronym commented:
That would be a good approach in delambdafy. But maybe harder in erasure. I
opted for the delegation approach as the path of least resistance.

gkossakowski commented:
Ok. I think we could give it a try when we work on making delambdafy the default. How about you add a TODO comment and fix a typo in the test that @Nami-Doc pointed out?

gkossakowski commented:
@retronym: ping

retronym committed:
SI-6260 Adddress pull request review

  - fix typo
  - remove BRIDGE flag from the method that we promote from
    a bridge to a bona-fide method
  - note possibility for delambdafy to avoid the bridge method
    creation in *all* cases.
  - note inconsistency with anonymous class naming between
    `-Ydelamdafy:{inline,method}`

retronym commented:
@gkossakowski fixed. I've also removed the BRIDGE flag, and noted an inconsistency in the way `Delambdafy` names the anonymous classes.

gkossakowski commented:
PLS REBUILD/pr-scala@d6b1e6e

scala-jenkins commented:
(kitty-note-to-self: ignore 34621292)
:cat: Roger! Rebuilding pr-scala for d6b1e6e4. :rotating_light:

gkossakowski commented:
LGTM.
